pennylane_qiskit/qiskit_device2.py

```diff
@@ -0,0 +1,606 @@
+
+        option_kwargs.pop("shots")
+        kwargs = self._init_kwargs.copy()
+        kwargs.update(option_kwargs)
+
+        self._kwargs = kwargs
+
+    @staticmethod
+    def get_transpile_args(kwargs):
+        """The transpile argument setter.
+
+        Keyword Args:
+            kwargs (dict): keyword arguments to be set for the Qiskit transpiler. For more details, see the
+                `Qiskit transpiler documentation <https://qiskit.org/documentation/stubs/qiskit.compiler.transpile.html>`_
+        """
+
+        transpile_sig = inspect.signature(transpile).parameters
+
+        transpile_args = {arg: kwargs[arg] for arg in transpile_sig if arg in kwargs}
+        transpile_args.pop("circuits", None)
+        transpile_args.pop("backend", None)
+
+        return transpile_args
+
+    def compile_circuits(self, circuits):
+        r"""Compiles multiple circuits one after the other.
+
+        Args:
+            circuits (list[QuantumCircuit]): the circuits to be compiled
+
+        Returns:
+             list[QuantumCircuit]: the list of compiled circuits
+        """
+        # Compile each circuit object
+        compiled_circuits = []
+        transpile_args = self.get_transpile_args(self._kwargs)
+
+        for circuit in circuits:
+            compiled_circ = transpile(circuit, backend=self.backend, **transpile_args)
+            compiled_circ.name = f"circ{len(compiled_circuits)}"
+            compiled_circuits.append(compiled_circ)
+
+        return compiled_circuits
+
+    # pylint: disable=unused-argument
+    def execute(
+        self,
+        circuits: QuantumTape_or_Batch,
+        execution_config: ExecutionConfig = DefaultExecutionConfig,
+    ) -> Result_or_ResultBatch:
+        session = self._session or Session(backend=self.backend)
+
+        if not self._use_primitives:
+            results = self._execute_runtime_service(circuits, session=session)
+            return results
+
+        results = []
+
+        if isinstance(circuits, (QuantumTape, QuantumScript)):
+            circuits = [circuits]
+
+        for circ in circuits:
+            if isinstance(circ.measurements[0], (ExpectationMP, VarianceMP)):
+                execute_fn = self._execute_estimator
+            elif isinstance(circ.measurements[0], ProbabilityMP):
+                execute_fn = self._execute_sampler
+            else:
+                execute_fn = self._execute_runtime_service
+            results.append(execute_fn(circ, session))
+
+        if self._session is None:
+            # if this was not a session set on the device, but just one created for this execution, close
+            session.close()
+
+        return results
+
+    def _execute_runtime_service(self, circuits, session):
+        """Execution using old runtime_service (can't use runtime sessions)"""
+
+        # update kwargs in case Options has been modified since last execution
+        self._update_kwargs()
```

Comment:
I second this question. and if it is needed, can we move the `shots` warning from `__init__` into that helper function? would rather not have it tracked twice, especially because there is a magic number in there

Reply:
> Is there ever a need for changing device configurations, or are all options locked in at device creation? If everything is locked in, why do we still allow users to edit configurations through the options object?

Kwargs are locked in at initialization and always have been. The `Options` object is mutable and can be updated between executions, which is why is here. Long term I guess the idea is we move away from the kwargs and only have the mutable options, but we won't remove this functionality until Qiskit removes the functionality it accesses (so possibly never).

> if it is needed, can we move the shots warning from __init__ into that helper function?
Yes, good point, there's no need to do this twice

Reply:
Here I fixed this with some assumptions.

The assumption is that whenever the options.executions.shots does not equal to the dev.shots we will always assume that this was because the User either 1) Edited the options.executions.shots or 2) Set the options.execution.shots at initialization. The error is raised when update_kwargs is run and it simply raises the error if the dev.shots and the options.shots are different (and then sets the options to match the dev). 

The second assumption is that the user did not purposefully choose 4000 aka write options.executions.shots = 4000 when they were defining options. If they did do so, they would not get a warning that says that shots cannot be defined through Options. This is a bit challenging to fix, but the alternative would be to warn the user that they set a common default value in Qiskit in the Options() object.

E.g. dev = QiskitDevice2(wires=2, backend=backend, options=Options())
Warning: Default number of shots (4000) detected in Options. If this was intentional, note that Pennylane does not support setting shots through Options. Device shots are set to 1024 {dev.shots}

In fact, we wouldn't need this warning if it was set this way:
E.g. dev = QiskitDevice2(wires = 2, backend=backend), since in the initialization we would see that options = None and know that the 4000 shots is a default value.